src/main.rs

```diff
@@ -27,37 +36,76 @@ fn main() {
     };
 
-    let ignore = Ignore::from_working_dir(&local_dir_absolute_path);
+    let ignore = Ignore::from_working_dir(&local_dir_absolute_path.clone());
 
-    let start = Instant::now();
+    println!("Pushing...");
 
-    if let Err(error) = sync_before_remote_command(&local_dir_absolute_path, &config, &ignore) {
-        exit_with_error(&format!("Sync local → remote machine failed: {}.", error), 1)
+    match sync::push(&local_dir_absolute_path, &config, &ignore) {
+        Err(err) => exit_with_error(&format!("Push failed: {}, took {}", err.message, format_duration(err.duration)), 1),
+        Ok(ok) => println!("Push done: took {}.\n", format_duration(ok.duration)),
     }
 
-    let remote_command_result = execute_remote_command(&local_dir_absolute_path, &args, &config);
+    match config.pull.mode {
+        PullMode::Serial => println!("Executing command on remote machine...\n"),
+        PullMode::Parallel(_) => println!("Executing command on remote machine (pulling in parallel)...\n")
+    }
+
+    let mut remote_command_readers = remote_command::execute_remote_command(
+        args.command.clone(),
+        config.clone(),
+        sync::project_dir_on_remote_machine(&local_dir_absolute_path.clone()),
+        2
+    );
+
+    let pull_finished_rx = sync::pull(&local_dir_absolute_path, config.clone(), ignore, &config.pull.mode, remote_command_readers.pop().unwrap());
 
-    if let Err(error) = sync_after_remote_command(&local_dir_absolute_path, &config, &ignore) {
-        exit_with_error(&format!("Sync remote → local machine failed: {}.", error), 1)
+    let remote_command_result = remote_command_readers
+        .pop()
+        .unwrap()
+        .recv()
+        .unwrap();
+
+    match remote_command_result {
+        Err(ref err) => eprintln!("\nExecution failed: took {}.\nPulling...", format_duration(err.duration)),
+        Ok(ref ok) => println!("\nExecution done: took {}.\nPulling...", format_duration(ok.duration))
     }
 
-    let duration = start.elapsed();
+    let pull_result = pull_finished_rx
+        .recv()
+        .expect("Could not receive remote_to_local_sync_result");
+
+    let total_duration = total_start.elapsed();
+
+    match pull_result {
+        Err(ref err) => eprintln!("Pull failed: {}, took {}.", err.message, format_duration(err.duration)),
+        Ok(ref ok) => println!("Pull done: took {}", format_duration(ok.duration)),
+    }
 
     match remote_command_result {
-        Err(_) => exit_with_error(&format!("\nFailure: took {}.", format_duration(duration)), 1),
-        _ => println!("\nSuccess: took {}.", format_duration(duration))
+        Err(_) => {
+            match pull_result {
+                Err(_)=> exit_with_error(&format!("\nFailure: took {}.", format_duration(total_duration)), 1),
+                Ok(_) => exit_with_error(&format!("\nFailure: took {}.", format_duration(total_duration)), 1),
+            }
+        },
+        Ok(_) => {
+            match pull_result {
```

Comment:
Try matching on the tuple `(remote_command_result, pull_result)` to make the whole match more concise.